gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -1144,6 +1154,15 @@ private void cleanUp() {
       for (String dagId: dagIdstoClean) {
         cleanUpDag(dagId);
       }
+
+      for (Iterator<Map.Entry<String, Dag<JobExecutionPlan>>> iter = this.failedDags.entrySet().iterator(); iter.hasNext();) {
```

Comment:
This should work. But my only concern is that we are looping over all the failed dags to check if their clean up time has come at the end of every DagManager poll cycle. Should we do this in a separate cleaning thread in DagManager, which runs, say every 1 hour or so?

Reply:
The javadoc also says "Because this thread runs in a regular interval, we should avoid doing repetitive work inside it."

Reply:
Good point, moved it to a separate thread for retention that runs every 60 mins by default.